### DISCUSSION TITLE
exploring-the-dotnet-8-preview-short-circuit-routing/

### DISCUSSION THREAD

**Discussion by giscus** [andrewlock/blog-comments#392] [Announcements]
# exploring-the-dotnet-8-preview-short-circuit-routing/

In this post I describe the new short-circuit routing feature, how it differs from normal routing, discuss why it's useful, and look at how it's implemented

https://andrewlock.net/exploring-the-dotnet-8-preview-short-circuit-routing/

**Comment by antonburger-alcove** [comment]
Totally tangential point in the implementation of `ExecuteShortCircuit` - just that I hadn't seen the technique used by `AwaitRequestTask` before, where the outer method avoids being `async` and all that comes with it, but still has the ability to return an awaited task if needed via a nested `async` function.

I guess similar in spirit to having a `IAsyncEnumerable<T>`-returning (non-iterator) method which delegates to a nested iterator. Though in that case, I use that in order to enable the use of `[EnumeratorCancellation]` without "polluting" the outer method's signature with an explicit `CancellationToken`.

**Comment by HakamFostok** [comment]
Thank you Andrew, that was a great post, I have a question and request.

Question: Is the Health endpoint a good use for this feature?
Request: Will you write a post about the Request Timeout feature that came in .NET 8?

Thank you very much, really appreciate it

**Comment by Porozhniakov** [comment]
Hello, Andrew! Thank you for the post.
I would like to add a small comment to the picture, which shows the order of the middleware.
Looking at this picture, the reader might think that this order of middleware is constant in the pipeline. In fact, this order will only exist if we add calls `.UseStaticFiles()` and `.UseRouting()` in exactly this order. Unless you call a `.UseRouting()`, as most projects do, the routing middleware will always be at the beginning of the pipeline.
I think the correct diagram for most projects that don't have a `.UseRouting()` and `.UseEndpoints()` calls would look like this:
Routing middleware <-> All other middleware like static files, CORS, logging, authorization, custom middleware etc. <-> Endpoint middleware

I also found a typo in the first paragraph of the "Taking a peak at the implementation" section: the first bullet point contains "ShorCircuit()" instead of "Shor**t**Circuit()".
